test/session.test.js

```diff
@@ -0,0 +1,160 @@
+/* eslint-disable no-unused-expressions */
+/* eslint-disable no-undef */
+import * as Factory from './lib/factory.js';
+chai.use(chaiAsPromised);
+const expect = chai.expect;
+
+describe('server session', () => {
+  const BASE_ITEM_COUNT = 1; /** Default items key */
+
+  const syncOptions = {
+    checkIntegrity: true,
+    awaitAll: true
+  };
+
+  before(async function () {
+    localStorage.clear();
+  });
+
+  after(async function () {
+    localStorage.clear();
+  });
+
+  beforeEach(async function () {
+    this.timeout(Factory.TestTimeout);
+    this.expectedItemCount = BASE_ITEM_COUNT;
+    this.application = await Factory.createInitAppWithRandNamespace();
+    this.email = Uuid.GenerateUuidSynchronously();
+    this.password = Uuid.GenerateUuidSynchronously();
+    await Factory.registerUserToApplication({
+      application: this.application,
+      email: this.email,
+      password: this.password
+    });
+  });
+
+  afterEach(async function () {
+    expect(this.application.syncService.isOutOfSync()).to.equal(false);
+    const items = this.application.itemManager.items;
+    expect(items.length).to.equal(this.expectedItemCount);
+    const rawPayloads = await this.application.storageService.getAllRawPayloads();
+    expect(rawPayloads.length).to.equal(this.expectedItemCount);
+    await this.application.deinit();
+  });
+
+  function getDelayBeforeNextRequest(currentSession) {
+    const tokenExpireAt = currentSession.expireAt;
+    const timeRemaining = Math.ceil(tokenExpireAt - Date.now() / 1000);
+    /* 
+      If the token has not expired yet, we will return the remaining time, plus one second.
+      Else, there's no need to add a delay.
+    */
+    return timeRemaining > 0 ? timeRemaining + 1 : 0;
+  }
+
+  async function getSessionFromStorage(application) {
+    return application.storageService.getValue(StorageKey.Session);
+  }
+
+  it('should fail when a sync request is perfomed with an expired access token', async function () {
+    const currentSession = this.application.apiService.session;
+
+    const delayBeforeNextRequest = getDelayBeforeNextRequest(currentSession);
+    await Factory.sleep(delayBeforeNextRequest);
+
+    const response = await this.application.apiService.sync([]);
+
+    expect(response).to.have.property('status');
+    expect(response.status).to.equal(498);
+
+    expect(response).to.have.property('error');
+
+    expect(response.error).to.have.property('tag');
+    expect(response.error.tag).to.equal('expired-access-token');
+
+    expect(response.error).to.have.property('message');
+    expect(response.error.message).to.equal('The provided access token has expired.');
+  }).timeout(10000);
+
+  it('should return the new session in the response when refreshed', async function () {
+    const response = await this.application.apiService.refreshSession();
+
+    expect(response).to.have.property('status');
```

Comment:
So the standard for how tests should be written should be modeled around newer tests that I've written. Older tests I've written tend to violate the principle of "test only one thing" per test. Here's a newer test for example: https://github.com/standardnotes/snjs/blob/004/test/item_manager.test.js#L120

So in this case you want to create a separate test for ensuring `status` exists. Then, you don't need to check this again for every test.

In this case you have two tests that both check for the same properties, when one test should do.

Reply:
(I don't necessarily mean you should just have one `expect` call per test, but just that we shouldn't repeat expect calls across multiple tests. And generally the less expect calls per test, the better)

Reply:
In this case, `status` is a property that is returned in the response. I'm checking that it has an specific value when the session is refreshed successfully. I will need to repeat this `expect` call on multiple tests. Perhaps I should access the property directly, instead of checking that it exists first? It should look like:

```javascript
it('should return the new session in the response when refreshed', async function () {
    const response = await this.application.apiService.refreshSession();

    expect(response.status).to.equal(200);
    expect(response.token).to.be.a('string');
    expect(response.token).to.not.be.empty;
    expect(response.session.expire_at).to.be.a('number');
    expect(response.session.refresh_token).to.not.be.empty;
  }).timeout(10000);
```